src/game/logic/ai/turretai.cpp

```diff
@@ -13,3 +13,651 @@
+                    }
+                } else {
+                    return STATE_CONTINUE;
+                }
+            }
+        }
+    }
+
+    if (turret->Is_Unk()) {
+        turret->Set_Turret_Target_Object(nullptr, false);
+    }
+
+    return STATE_FAILURE;
+}
+
+void TurretAIAimTurretState::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char version = 1;
+    xfer->xferVersion(&version, 1);
+}
+
+TurretAIRecenterTurretState::TurretAIRecenterTurretState(TurretStateMachine *machine) :
+    TurretState(machine, "TurretAIRecenterTurretState")
+{
+}
+
+StateReturnType TurretAIRecenterTurretState::Update()
+{
+    if (Get_Machine_Owner()->Get_Status(OBJECT_STATUS_UNDER_CONSTRUCTION)) {
+        return STATE_CONTINUE;
+    }
+
+    TurretAI *turret = Get_Turret_AI();
+    bool angle = turret->Friend_Turn_Towards_Angle(turret->Get_Natural_Turret_Angle(), 0.5f, 0.0f);
+    bool pitch = turret->Friend_Turn_Towards_Pitch(turret->Get_Natural_Turret_Pitch(), 0.5f);
+
+    if (angle && pitch) {
+        return STATE_SUCCESS;
+    } else {
+        return STATE_CONTINUE;
+    }
+}
+
+void TurretAIRecenterTurretState::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char version = 1;
+    xfer->xferVersion(&version, 1);
+}
+
+TurretAIIdleState::TurretAIIdleState(TurretStateMachine *machine) :
+    TurretState(machine, "TurretAIIdleState"), m_nextIdleScan(0)
+{
+}
+
+StateReturnType TurretAIIdleState::On_Enter()
+{
+    AIUpdateInterface *update = Get_Machine_Owner()->Get_AI_Update_Interface();
+
+    if (update != nullptr) {
+        update->Reset_Next_Mood_Check_Time();
+
+        if (update->Get_Current_Turret() == Get_Turret_AI()->Get_Which_Turret()) {
+            update->Set_Current_Turret(TURRET_INVALID);
+        }
+    }
+
+    Reset_Idle_Scan();
+
+    return Frame_To_Sleep_Time(
+        Get_Turret_AI()->Friend_Get_Next_Idle_Mood_Target_Frame(), m_nextIdleScan, 0x3FFFFFFF, 0x3FFFFFFF);
+}
+
+StateReturnType TurretAIIdleState::Update()
+{
+    if (g_theGameLogic->Get_Frame() >= m_nextIdleScan) {
+        return STATE_FAILURE;
+    }
+
+    TurretAI *turret = Get_Turret_AI();
+    turret->Friend_Check_For_Idle_Mood_Target();
+    return Frame_To_Sleep_Time(turret->Friend_Get_Next_Idle_Mood_Target_Frame(), m_nextIdleScan, 0x3FFFFFFF, 0x3FFFFFFF);
+}
+
+void TurretAIIdleState::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char version = 1;
+    xfer->xferVersion(&version, 1);
+    xfer->xferUnsignedInt(&m_nextIdleScan);
+}
+
+void TurretAIIdleState::Reset_Idle_Scan()
+{
+    int frame = g_theGameLogic->Get_Frame();
+    unsigned int max = Get_Turret_AI()->Get_Max_Idle_Scan_Interval();
+    unsigned int min = Get_Turret_AI()->Get_Min_Idle_Scan_Interval();
+    m_nextIdleScan = Get_Logic_Random_Value(min, max);
+}
+
+TurretAIIdleScanState::TurretAIIdleScanState(TurretStateMachine *machine) :
+    TurretState(machine, "TurretAIIdleScanState"), m_desiredAngle(0.0f)
+{
+}
+
+StateReturnType TurretAIIdleScanState::On_Enter()
```

Comment:
Should this perhaps return `UpdateSleepTime` here?

Reply:
No, this definitely returns StateReturnType.